### PR TITLE
David.benque/stateless controller

### PR DESCRIPTION
**Pull request by dbenque**

### CODE REVIEW

File: cmd/draino/draino.go [hunk 1]
```diff
@@ -142,33 +147,38 @@ func main() {
...
 	kingpin.FatalIfError(err, "cannot create lock")
 
-	leaderelection.RunOrDie(leaderelection.LeaderElectionConfig{
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
 		Lock:          lock,
 		LeaseDuration: 15 * time.Second,
 		RenewDeadline: 10 * time.Second,
 		RetryPeriod:   2 * time.Second,
 		Callbacks: leaderelection.LeaderCallbacks{
-			OnStartedLeading: func(_ <-chan struct{}) {
```
Comment (CharlyF) [inline cmd/draino/draino.go]: One of the available callbacks is `OnNewLeader`, maybe it could be leveraged ?

File: cmd/draino/draino.go [hunk 2]
```diff
@@ -213,11 +223,3 @@ func (r *httpRunner) Run(stop <-chan struct{}) {
 	s.ListenAndServe() // nolint:gosec
 	cancel()
 }
-
-// Many Kubernetes client things depend on glog. glog gets sad when flag.Parse()
-// is not called before it tries to emit a log line. flag.Parse() fights with
-// kingpin.
-func glogWorkaround() {
```
Comment (CharlyF) [inline cmd/draino/draino.go]: Why is this removed?

File: internal/kubernetes/drainSchedule.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+	} else {
+		d.logger.Error("Failed schedule deletion", zap.String("key", name))
+	}
+	delete(d.schedules, name)
+}
+
+func (d *DrainSchedules) WhenNextSchedule() time.Time {
+	// compute drain schedule time
+	sooner := time.Now().Add(2 * TimeoutToSetCondition)
+	when := d.lastDrainScheduledFor.Add(d.period)
```
Comment (CharlyF) [inline internal/kubernetes/drainSchedule.go]: Dumb question, but this method mutates `lastDrainScheduledFor` and you re-affect `when` [later on](https://github.com/DataDog/draino/pull/2/files?file-filters%5B%5D=.go&file-filters%5B%5D=.md&file-filters%5B%5D=.mod&file-filters%5B%5D=.sum#diff-2b8ef4fa4610dda8146744e393cd3146R91).
I would expect the method to affect the right `when` to `lastDrainScheduledFor`. Is there any reason why it's not like that ?

File: internal/kubernetes/drainSchedule.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+		when = sooner
+	}
+	return when
+}
+
+func (d *DrainSchedules) Schedule(node *v1.Node) (time.Time, error) {
+	d.Lock()
+	if sched, ok := d.schedules[node.GetName()]; ok {
+		d.Unlock()
+		return sched.when, NewAlreadyScheduledError() // we already have a schedule planned
```
Comment (CharlyF) [inline internal/kubernetes/drainSchedule.go]: It might be later on in the code, but do we expire planned schedules ?
If a node has a schedule and for whatever reason it fails to schedule. We will never retry ?

File: internal/kubernetes/drainSchedule.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+	// Mark the node with the condition stating that drain is scheduled
+	if err := Until(
+		func() error {
+			return d.drainer.MarkDrain(node, when, time.Time{}, false)
+		},
+		50*time.Millisecond,
+		TimeoutToSetCondition,
+	); err != nil {
+		// if we cannot mark the node, let's remove the schedule
```
Comment (CharlyF) [inline internal/kubernetes/drainSchedule.go]: I think it'd be worth logging something, if we can't mark a schedule because of a timeout it could be because of an underlying networking problem or something. Maybe the node should be marked by the Node Problem Detector down the road ?

File: internal/kubernetes/drainSchedule.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+				},
+				50*time.Millisecond,
+				TimeoutToSetCondition,
+			); err != nil {
+				log.Error("Failed to place condition following drain failure")
+			}
+			return
+		}
+		sched.finish = time.Now()
+		log.Info("Drained")
```
Comment (CharlyF) [inline internal/kubernetes/drainSchedule.go]: Could we add the node name in the log ?

File: internal/kubernetes/drainer.go [hunk 7]
```diff
@@ -161,6 +170,68 @@ func (d *APICordonDrainer) Cordon(n *core.Node) error {
 	return nil
 }
 
+// MarkDrain set a condition on the node to mark that that drain is scheduled.
+func (d *APICordonDrainer) MarkDrain(n *core.Node, when, finish time.Time, failed bool) error {
+	nodeName := n.Name
+	// Refresh the node object
+	freshNode, err := d.c.CoreV1().Nodes().Get(nodeName, meta.GetOptions{})
```
Comment (CharlyF) [inline internal/kubernetes/drainer.go]: note to self: check if there is an informer or if the call is made to the APIServer

File: internal/kubernetes/drainer.go [hunk 8]
```diff
@@ -161,6 +170,68 @@ func (d *APICordonDrainer) Cordon(n *core.Node) error {
...
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
```
Comment (CharlyF) [inline internal/kubernetes/drainer.go]: why do you only return the error if it's not found ? If there is a timeout/unauthorized/conflict etc, it would be worth surfacing ?

Comment (CharlyF) [inline internal/kubernetes/drainer.go]: note to self check when MarkDrain is called and the node is not found

File: internal/kubernetes/nodefilters.go [hunk 9]
```diff
@@ -83,16 +83,6 @@ func ParseConditions(conditions []string) []SuppliedCondition {
 	return parsed
 }
 
-// NodeSchedulableFilter returns true if the supplied object is a schedulable
-// node.
-func NodeSchedulableFilter(o interface{}) bool {
```
Comment (CharlyF) [inline internal/kubernetes/nodefilters.go]: Why removing this method when it could be used https://github.com/DataDog/draino/pull/2/files#diff-86d89adaf5925d6ff41a251e1317875cR138

File: internal/kubernetes/drainSchedule.go (hunk 5 above)
Comment (CharlyF) [reply internal/kubernetes/drainSchedule.go]: I read the definition of `Until` after writing this comment. If there is a timeout it will be surfaced.

File: cmd/draino/draino.go (hunk 1 above)
Comment (dbenque) [reply cmd/draino/draino.go]: I did not want to touch other logic that then one linked to stateless-controller goal to maximize the chances to have an upstream merge.
This being said you are probably right, I haven't check this part and once the Pr will be merged upstream, I may change this.

File: cmd/draino/draino.go (hunk 2 above)
Comment (dbenque) [reply cmd/draino/draino.go]: I changed the go version and lots of dependencies.
The kubernetes client-go that I am now using is based on klog, the has fix many of the glog problems. Also that old workaround was simply breaking the program start-up.
I had the same question upstream:
https://github.com/planetlabs/draino/pull/67#discussion_r396591750

File: internal/kubernetes/drainSchedule.go (hunk 3 above)
Comment (dbenque) [reply internal/kubernetes/drainSchedule.go]: This method does not mutate `lastDrainScheduledFor`
https://play.golang.org/p/kkXTWh9hiRJ

The time.Add function is not on a pointer receiver https://golang.org/pkg/time/#Time.Add

File: internal/kubernetes/drainSchedule.go (hunk 4 above)
Comment (dbenque) [reply internal/kubernetes/drainSchedule.go]: The schedule does not expire. The schedule is removed in case of delete:
https://github.com/DataDog/draino/blob/b6514339b8a29dd99bbe2ea4cce01e86e3621b3f/internal/kubernetes/drainSchedule.go#L66

If the schedule is not delete it will lead to either a Completed or Failed drain activity and the finish time will be set:
https://github.com/DataDog/draino/blob/b6514339b8a29dd99bbe2ea4cce01e86e3621b3f/internal/kubernetes/drainSchedule.go#L115

In case of failure the user can annotate the node to have another schedule created automatically:
https://github.com/DataDog/draino/blob/b6514339b8a29dd99bbe2ea4cce01e86e3621b3f/internal/kubernetes/eventhandler.go#L152

File: internal/kubernetes/drainSchedule.go (hunk 5 above)
Comment (dbenque) [reply internal/kubernetes/drainSchedule.go]: Yes, also in the latest version I have replaced the Until by another function of the `wait` package that does the same thing:
https://github.com/DataDog/draino/blob/b6514339b8a29dd99bbe2ea4cce01e86e3621b3f/internal/kubernetes/util.go#L56

File: internal/kubernetes/drainer.go (hunk 7 above)
Comment (dbenque) [reply internal/kubernetes/drainer.go]: It is already done here:
https://github.com/DataDog/draino/blob/b6514339b8a29dd99bbe2ea4cce01e86e3621b3f/internal/kubernetes/drainSchedule.go#L124

File: internal/kubernetes/drainer.go (hunk 8 above)
Comment (dbenque) [reply internal/kubernetes/drainer.go]: I am not sure to get what you mean "self check".
My logic here is that if the node is gone, there is no problem, in a way it is already drained.
If a schedule is set prior to that MarkDrain activity the schedule will be remove because the delete node will be caught by the controller that will delete the schedule before it can be executed.

File: internal/kubernetes/nodefilters.go (hunk 9 above)
Comment (dbenque) [reply internal/kubernetes/nodefilters.go]: This method was more for filtering stack implementation:
https://github.com/DataDog/draino/blob/b6514339b8a29dd99bbe2ea4cce01e86e3621b3f/cmd/draino/draino.go#L161

In the other place, the code that check directly the unschedule field is far more efficient (and small), there is not type checking to do.

File: internal/kubernetes/drainSchedule.go (hunk 6 above)
Comment (dbenque) [reply internal/kubernetes/drainSchedule.go]: It is already done here:
https://github.com/DataDog/draino/blob/b6514339b8a29dd99bbe2ea4cce01e86e3621b3f/internal/kubernetes/drainSchedule.go#L124

File: internal/kubernetes/drainer.go (hunk 8 above)
Comment (CharlyF) [reply internal/kubernetes/drainer.go:182]: I was missing a `:` or `,`.
I wanted to check what happens when MarkDrain is called and the node is not found.
As you said:
>  in a way it is already drained
So it's all good.
